File: 82.remove-duplicates-from-sorted-list-ii.java
/*
 * @lc app=leetcode id=82 lang=java
 *
 * [82] Remove Duplicates from Sorted List II
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode be4Head = new ListNode(0);
        be4Head.next = head;
        ListNode prev = be4Head, ptr = head, nextPtr = ptr.next;
        boolean initSaveVal = false;
        int savedVal = 0;
        while (nextPtr != null) {
            if (ptr.val == nextPtr.val || (initSaveVal && ptr.val == savedVal)) {
                initSaveVal = true;
                savedVal = ptr.val;
                prev.next = nextPtr;
            } else {
                prev = ptr;
            }
            ptr = nextPtr;
            nextPtr = nextPtr.next;
        }
        if (initSaveVal && ptr.val == savedVal) prev.next = null;
        return be4Head.next;
    }
}

File: go/82.remove-duplicates-from-sorted-list-ii.go
/*
 * @lc app=leetcode id=82 lang=golang
 *
 * [82] Remove Duplicates from Sorted List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
    if head == nil || head.Next == nil {
		return head
	}
	ahead := &ListNode{0, head}
	prev, cur := ahead, head
	for cur != nil {
		if cur.Next == nil || cur.Next.Val != cur.Val {
			prev, cur = cur, cur.Next
		} else {
			val := cur.Val
			for cur != nil && cur.Val == val {
				cur = cur.Next
			}
			prev.Next = cur
		}
	}
	return ahead.Next
}
// @lc code=end

